Multi-material solid mechanics

btalamini opened the pull request:
Enable solid mechanics to work with materials defined on different domains. Includes materials with internal state variables.

For now, I added this without breaking changes. This means I overloaded the `setMaterial` several times and left in place the old ones that do not take a domain. Eventually, I'd prefer the domain to be required. We can map this out later.

btalamini committed:
Fix a missing const in Functional method

btalamini committed:
Implement multi-material solid mechanics, test by eye

btalamini committed:
Make test rigorous

btalamini committed:
Make domain optional, default to whole mesh

btalamini committed:
Handle internal variables (quadrature data)

btalamini committed:
FIll in missing values for geometry type counts

btalamini committed:
Ipmrpove general comments of tests

btalamini committed:
Add another setMaterial overload to avoid breaking changes for anyone

btalamini committed:
Merge branch 'develop' into talamini/feature/multi_material

btalamini committed:
Make style

btalamini committed:
Remove dead code

btalamini committed:
Fix broken doxygen comment

btalamini committed:
Add missing doxygen

btalamini committed:
Add yet another overload of setMaterial to avoid breaking changes

btalamini committed:
Make style

btalamini committed:
Add explicit template arguments to fix errors on ancient clang

white238 commented on src/serac/physics/state/state_manager.hpp:
```diff
@@ -130,6 +130,39 @@ class StateManager {
+  {
+    int Q = order + 1;
+
+    std::array<uint32_t, mfem::Geometry::NUM_GEOMETRIES> elems = geometry_counts(domain);
```
There is only two lines different in this vs the `mesh_tag` `newQuadratureDataBuffer` function. Can you reduce the duplication between them?

btalamini replied:
Yes. I removed the `mesh_tag` version. It wasn't being used. The solid mechanics module handles the logic of whether the variables are allocated over the entire mesh or a single `Domain`.

white238 commented on src/serac/physics/tests/solid_multi_material.cpp:
```diff
@@ -0,0 +1,280 @@
+
+  auto& pmesh = serac::StateManager::setMesh(std::move(mesh), mesh_tag);
+
+  // _solver_params_start
```
These tags are used in the sphinx documentation and are not needed here. Looking at the code these tags have bled into other files and should only be in `serac/physics/tests/solid.cpp`.

btalamini replied:
Good catch, I didn't even see those.

white238 commented on src/serac/physics/tests/solid_multi_material.cpp:
```diff
@@ -0,0 +1,280 @@
+
+#include "serac/physics/solid_mechanics.hpp"
+
+#include "axom/slic/core/SimpleLogger.hpp"
```
```suggestion
```
This isn't needed due to using the `serac::initiliaze` call which uses the Serac SLIC logger.

btalamini replied:
Done

samuelpmishLLNL commented on src/serac/numerics/functional/functional.hpp:
```diff
@@ -301,7 +301,7 @@ class Functional<test(trials...), exec> {
   /// @overload
   template <int dim, int... args, typename lambda, typename qpt_data_type = Nothing>
-  void AddDomainIntegral(Dimension<dim>, DependsOn<args...>, const lambda& integrand, Domain& domain,
+  void AddDomainIntegral(Dimension<dim>, DependsOn<args...>, const lambda& integrand, const Domain& domain,
```
the DG PR requires making changes to the domain class when defining integrals, so we might have to revert this back to non-const `Domain & domain` soon (or use `mutable` stuff)

btalamini replied:
Noted. There's the usual virality: If we take away the `const` here, it will have to propagate down to `Functional` (which is why I added it here). I'll let you pick which.

btalamini committed:
Remove redundant allocation function for quadrature data

I think it's better to force an explicit domain. We don't need
an overloaded versino for the whole mesh, since we have the
EntireDomain(mesh) function.

btalamini committed:
Address Chris's other comments

btalamini committed:
Merge branch 'develop' into talamini/feature/multi_material
